spec/src/main/asciidoc/property-expansion.asciidoc

```diff
@@ -0,0 +1,70 @@
+//
+// Copyright (c) 2020 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// You may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// Contributors:
+// Roberto Cortez
+
+[[property-expansion]]
+== Property Expansion
+
+The value of a configuration property may contain an expression corresponding to another configuration property. An
+expression string is a mix of plain strings and expression segments, which are wrapped by the sequence `${ ... }`.
+
+Consider the following configuration properties file:
+
+[source,properties]
+----
+server.url=http://${server.host}/endpoint
+server.host=example.org
+----
+
+When looking up the `server.url` property, the value will be resolved and expanded to `http://example.org/endpoint`.
+All MicroProfile Config rules still apply. The `Config` is able to resolve expressions from different ConfigSources.
+
+Additionally, it is also possible to use the following syntax with for property expansion:
+
+* `${expression:value}` - Provides a default value after the `:` if the expansion doesn't find a value.
+* `${my.prop${compose}}` - Composed expressions. Inner expressions are resolved first.
```

Comment:
We need to be careful to support the nested variable replacement. How deep do we want to go? 

The implementation must detect the infinite loop and exit with IllegalArgumentException. See the example below.

a=${${b}}
b=${a}
c=${d}
d=${c}

Reply:
Http spec has this stated:
10.3 Redirection 3xx

   This class of status code indicates that further action needs to be
   taken by the user agent in order to fulfill the request.  The action
   required MAY be carried out by the user agent without interaction
   with the user if and only if the method used in the second request is
   GET or HEAD. **A client SHOULD detect infinite redirection loops, since
   such loops generate network traffic for each redirection.**

      Note: previous versions of this specification recommended a
      maximum of **five** redirections. Content developers should be aware
      that there might be clients that implement such a fixed
      limitation.

Though it is not exactly same, it is quite close by nature. I am wondering whether we could borrow the similar approach.